### PR TITLE
[eas-cli] Improve configuration setup for EAS update

### PR DESCRIPTION
**Pull request by byCedric**
# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# TODO

- Add tests for `src/updates/configure.ts` -> comming in follow up PR

# Why

- Continuation of PR https://github.com/expo/eas-cli/pull/1473 and [ENG-5926](https://linear.app/expo/issue/ENG-5926)
- Fixes ENG-6748, ENG-5806
- Part of ENG-6002

# How

This moves most of the (auto) configuration from both `eas update:configure` and `eas update` into `src/updates/configure.ts` and reuses it on both commands.

> This also includes the change from PR #1467, just for completeness and being able to test the final result with `npx @bycedric/some-cli-test update` - a published version of these stacked PRs.

# Result

## `eas update:configure`

<details><summary>Project without any EAS service/config setup</summary>

image

</details>

<details><summary>Project with EAS Update fully configured</summary>

image

</details>

<details><summary>Project with EAS Update fully configured, except <code>runtimeVersion</code></summary>

image

</details>

## `eas update:configure --non-interactive`

Adding this flag is required to ensure people can set their project up without `extra.eas.projects`. When always forcing non interactive, people will be prompted with the first example here:

<details><summary>Project without any EAS service/config setup</summary>

image

</details>

<details><summary>Project with just <code>extra.eas.projectId</code> setup</summary>

image

</details>

## `eas update --auto`

<details><summary>Project without any EAS service/config setup</summary>

image

</details>

<details><summary>Project with just <code>extra.eas.projectId</code> setup</summary>

image

</details>

<details><summary>Project with EAS Update fully configured</summary>

image

</details>

<details><summary>Project with EAS Update fully configured, except <code>runtimeVersion</code></summary>

image

</details>

# Test Plan

You can use my "fork package" on npm to play with the current state. It's `@bycedric/some-cli-test@3.0.4`, e.g. `npx @bycedric/some-cli-test@3.0.4 update` in a brand new project.

- `eas update:configure`
- `eas update`

### CODE REVIEW

**Comment by kbrandwijk** [comment]
At first glance, this seems to be somewhat incompatible with https://github.com/expo/eas-cli/pull/1427. Did you consider that PR?

**Comment by byCedric** [comment]
> to be somewhat incompatible with #1427. Did you consider that PR?

Yes, that PR has a different intent in what we want `eas update` to be. Evan and I created the version from this PR as a prototype to help remove friction encountered when (starting to) use EAS Updates. The friction here being mostly:
1. Having to run `eas init`, `eas update:configure`, and `eas update`.
2. Running into multiple errors, some of which are not helpful

Basically, whenever someone runs `eas update`, we assume intent that this user wants to use EAS Updates. With that in mind, we automate most of the setup while providing the user enough context to know that just running that command won't be sufficient. They also need to rebuild their app to start using EAS Updates or open it in Expo Go.

I do like some of the refactors that @Simek brought in. I already told @jonsamp that it's great to merge that PR before this one. I can rebase upon that one. But these stacked PRs take it a bit further, like dropping some exported functions from classes and reusing the same code for both `eas update:configure` and `eas update`.

**Comment by byCedric** [comment]
In the prototype version, I also started refactoring the `eas update` runAsync method, but I kept that out of these PRs to wait until both PR #1427 and #1467 are merged.

**Comment by byCedric** [comment]
/changelog-entry chore Remove friction from the initial EAS update setup

File: packages/eas-cli/src/update/configure.ts
Comment (EvanBacon) [inline packages/eas-cli/src/update/configure.ts:107]:
```suggestion
    runtimeVersion: modifyConfig.runtimeVersion ?? exp.runtimeVersion,
    updates: { ...exp.updates, ...modifyConfig.updates },
    android: {
      ...exp.android,
      ...modifyConfig.android
    },
    ios: {
      ...exp.ios,
      ...modifyConfig.ios
    },
```
This should work better

File: packages/eas-cli/src/update/configure.ts
Comment (EvanBacon) [inline packages/eas-cli/src/update/configure.ts:115]:
```suggestion
```

If we drop the `...exp` then we won't need the naive clean-up phase.

File: packages/eas-cli/src/update/configure.ts
Comment (EvanBacon) [inline packages/eas-cli/src/update/configure.ts:156]: not necessary but this block and the next (ios, android) should be a loop.

Comment (byCedric) [reply packages/eas-cli/src/update/configure.ts:149]: Yeah, to do this, we need to simplify the types. Right now `appPlatformDisplayNames` uses `AppPlatform` (which is all caps `ANDROID`/`IOS`). But, `ExpoConfig` uses lowercase `Platform`.

So to minimize the type of juggling, I kept it the same.

**Review by EvanBacon — APPROVED** [review]

File: packages/eas-cli/src/commands/update/configure.ts
Comment (wschurman) [inline packages/eas-cli/src/commands/update/configure.ts:139]: looks like the new version of this doesn't have this in the manual instruction?

File: packages/eas-cli/src/commands/update/configure.ts
Comment (wschurman) [inline packages/eas-cli/src/commands/update/configure.ts:120]: would be good to add a unit test for just this piece of code that works before and after your refactor

File: packages/eas-cli/src/update/configure.ts
Comment (wschurman) [inline packages/eas-cli/src/update/configure.ts:191]:
```suggestion
      `You will also have to manually edit the project's ${chalk.bold(
```

File: packages/eas-cli/src/update/configure.ts
Comment (wschurman) [inline packages/eas-cli/src/update/configure.ts:272]: is there a reason we wouldn't want to always do this (like the configure command used to do)?

**Review by wschurman — CHANGES_REQUESTED** [review]
The main thing we should make sure of is that the dynamic config stuff continues to work as we want it to (not saying that the way it worked previously was best, but if we want to change its behavior we should do so intentionally).

File: packages/eas-cli/src/update/configure.ts
Comment (EvanBacon) [inline packages/eas-cli/src/update/configure.ts:67]: Not super important but we had a lot of this kinda logic in `expo export` and `expo publish` where everything was duplicated twice. Prefer loops even though there are only two values.

File: packages/eas-cli/src/update/configure.ts
Comment (EvanBacon) [inline packages/eas-cli/src/update/configure.ts:107]: Would be cool if this was split into a different function that returned `mergedExp` so you could easily test the results and ensure the clobbering always works as expected. I could see a situation where someone changes the `modifyConfig` to define a value like `runtimeVersion: undefined` which then removes the `runtimeVersion` altogether.

File: packages/eas-cli/src/update/configure.ts
Comment (EvanBacon) [inline packages/eas-cli/src/update/configure.ts:182]:
```suggestion
    `Finish configuring your project for EAS Update by adding the following to the project app.config.js:\n${learnMore(
```

File: packages/eas-cli/src/update/configure.ts
Comment (EvanBacon) [inline packages/eas-cli/src/update/configure.ts:183]:
```suggestion
      'https://expo.fyi/eas-update-config'
```

File: packages/eas-cli/src/update/configure.ts
Comment (EvanBacon) [inline packages/eas-cli/src/update/configure.ts:193]:
```suggestion
      chalk`The native config files {bold Expo.plist & AndroidManifest.xml} must be updated to support EAS Update. ${learnMore('https://expo.fyi/eas-update-config.md#native-configuration')}`
```

File: packages/eas-cli/src/update/configure.ts
Comment (EvanBacon) [inline packages/eas-cli/src/update/configure.ts:258]: [sanity] Toggle off silent when debug logging is enabled, makes our job easier in the future.

File: packages/eas-cli/src/update/configure.ts
Comment (byCedric) [reply packages/eas-cli/src/update/configure.ts:272]: The main reason for that is to keep `eas update` performance similar to how it was before. We can run most of these `ensureX` commands, but it will slow it down.

That being said, haven't benchmarked it.

File: packages/eas-cli/src/commands/update/configure.ts
Comment (byCedric) [reply packages/eas-cli/src/commands/update/configure.ts:120]: Yeah, I like that too. Unfortunately, this method is one of those problematic methods that are really _really_ hard to test without setting up a project from scratch.

It does a lot of things:
  - Detecting if updates is set up in `app.json`
  - Detecting the current type of project (generic vs managed)
  - Make changes relative to the "requested platform" and current `runtimeVersion`
  - Make sure native files are synced relative to "requested platform" and the current type of project
  - Try to log the steps made, relative to the type of project

If we had to write tests for this, there would be so many cases and variables to cover everything. It's also the main reason why I split this out over multiple functions in this PR.

That being said, we can definitely add these tests for the new split functions without having to set up whole projects during testing.

File: packages/eas-cli/src/commands/update/configure.ts
Comment (byCedric) [reply packages/eas-cli/src/commands/update/configure.ts:139]: I think you are right. I can compensate for that in the `warnEASUpdatesManualConfig`. We can iterate over the `modifyConfig` and replace `undefined` with this string.

File: packages/eas-cli/src/update/configure.ts
Comment (byCedric) [reply packages/eas-cli/src/update/configure.ts:67]: One thing we can do here is changing the order of changes, much like how AST sanitization would work.

Instead of checking ahead of time if `runtimeVersions` are similar, we can do that after.
- Loop per platform first
  - Make required changes
- If requested platform === all && runtimeVersion is identical for all platforms: move it one level up

Does that sound better?

File: packages/eas-cli/src/update/configure.ts
Comment (byCedric) [reply packages/eas-cli/src/update/configure.ts:107]: Setting something to `undefined` does indeed remove that property, at least when using `JSON.parse` (there is no `undefined` type in JSON).

I'll split it regardless for future testing.

File: packages/eas-cli/src/update/configure.ts
Comment (byCedric) [reply packages/eas-cli/src/update/configure.ts:48]: Probably for testing later, but I'll remove it now.

File: packages/eas-cli/src/update/configure.ts
Comment (byCedric) [reply packages/eas-cli/src/update/configure.ts:203]: Probably for testing later, but I'll remove it now.

File: packages/eas-cli/src/commands/update/configure.ts
Comment (byCedric) [reply packages/eas-cli/src/commands/update/configure.ts:139]: Done

**Comment by byCedric** [comment]
@wschurman made the changes, going to merge now. Feel free to reach out if anything comes up

**Pull request merged**
